tests/helpers/test_aiohttp_client.py

```diff
@@ -91,6 +93,82 @@ async def test_get_clientsession_cleanup_without_ssl(hass):
     assert hass.data[client.DATA_CONNECTOR_NOTVERIFY].closed
 
 
+async def test_get_clientsession_patched_close(hass):
+    """Test closing clientsession does not work."""
+    with patch("aiohttp.ClientSession.close") as mock_close:
+        session = client.async_get_clientsession(hass)
+
+        assert isinstance(hass.data[client.DATA_CLIENTSESSION], aiohttp.ClientSession)
+        assert isinstance(hass.data[client.DATA_CONNECTOR], aiohttp.TCPConnector)
+
+        with pytest.raises(RuntimeError):
+            await session.close()
+
+        assert mock_close.call_count == 0
+
+
+async def test_warning_close_session_integration(hass, caplog):
+    """Test log warning message when closing the session from integration context."""
+    with patch(
+        "homeassistant.helpers.frame.extract_stack",
+        return_value=[
+            Mock(
+                filename="/home/paulus/homeassistant/core.py",
+                lineno="23",
+                line="do_something()",
+            ),
+            Mock(
+                filename="/home/paulus/homeassistant/components/hue/light.py",
+                lineno="23",
+                line="await session.close()",
+            ),
+            Mock(
+                filename="/home/paulus/aiohue/lights.py",
+                lineno="2",
+                line="something()",
+            ),
+        ],
+    ):
+        session = client.async_get_clientsession(hass)
+        await session.close()
+    assert (
+        "Detected integration that closes the Home Assistant aiohttp session. "
+        "Please report issue for hue using this method at "
+        "homeassistant/components/hue/light.py, line 23: await session.close()"
+    ) in caplog.text
+
+
+async def test_warning_close_session_custom(hass, caplog):
+    """Test log warning message when closing the session from custom context."""
+    with patch(
+        "homeassistant.helpers.frame.extract_stack",
+        return_value=[
+            Mock(
+                filename="/home/paulus/homeassistant/core.py",
+                lineno="23",
+                line="do_something()",
+            ),
+            Mock(
+                filename="/home/paulus/config/custom_components/hue/light.py",
+                lineno="23",
+                line="await session.close()",
+            ),
+            Mock(
+                filename="/home/paulus/aiohue/lights.py",
+                lineno="2",
+                line="something()",
+            ),
+        ],
+    ):
+        session = client.async_get_clientsession(hass)
+        await session.close()
+    assert (
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5518609)